### ISSUE TITLE
Questions on shmem_ctx_t

### ISSUE THREAD

**Issue by nspark** [openshmem-org/specification#221]
I'm starting to work with OpenSHMEM 1.4 and the new OpenSHMEM contexts features, but I have some questions on `shmem_ctx_t` that I wonder whether we need to answer/clarify in the specification.

Consider the following code, which is intended to be similar to the thread-registration extensions in Cray SHMEM:

```c
// Assume C11 support
_Thread_local shmem_ctx_t thread_ctx;

void mylib_thread_register(void) {
  shmem_ctx_create(SHMEM_CTX_PRIVATE, &thread_ctx);
}

void mylib_thread_unregister(void) {
  shmem_ctx_destroy(thread_ctx);
}

inline void mylib_putmem(void *dst, const void *src, size_t size, int pe) {
  shmem_putmem(thread_ctx, dst, src, size, pe);
}
```

### Default initializers

Currently, there is no compile-time constant (e.g., `SHMEM_CTX_NULL` or `SHMEM_CTX_INVALID`) to which `thread_ctx` can be initialized. Without such a feature, it seems difficult to make calls to `mylib_thread_register()` idempotent. For example, if I had such a feature, I would write:

```c
_Thread_local shmem_ctx_t thread_ctx = SHMEM_CTX_INVALID;
void mylib_thread_register(void) {
  if (thread_ctx == SHMEM_CTX_INVALID)
    shmem_ctx_create(SHMEM_CTX_PRIVATE, &thread_ctx);
}
```

Without such a feature, I cannot test whether `thread_ctx` has already been created by the calling thread.

### Safety of `shmem_ctx_destroy`

Such a feature also allows one to more safely indicate that a `shmem_ctx_t` object has been destroyed with an idiom like:

```c
void mylib_thread_unregister(void) {
  shmem_ctx_destroy(thread_ctx);
  thread_ctx = SHMEM_THREAD_INVALID;
}
```

In Cray SHMEM's extensions, a call to `shmemx_thread_unregister()` from an unregistered thread has no effect. I do not think that repeated calls to `shmem_ctx_destroy` with the same `shmem_ctx_t` object should be valid (much like successive`free()` calls with the same pointer are an error), however, a call of `shmem_ctx_destroy(SHMEM_CTX_INVALID)` could (and should, I think) be valid (much like `free(NULL)`).

### Miscellany

* Because `SHMEM_CTX_DEFAULT` is a link-time constant and not a compile-time constant, it cannot be used as an initializer for a global `shmem_ctx_t` object.
* In Sandia OpenSHMEM, which I'm using for this initial development, the `shmem_ctx_t` type(def) is a pointer type. I could static initialize it to `NULL`, however, that would not be portable, as OpenSHMEM 1.4 does not require `shmem_ctx_t` to be a pointer type.
* IIRC, @jdinan didn't want a constant like `SHMEM_CTX_INVALID` because he didn't want the expectation that the library would test and print a sensible error in the case that an invalid context was used (e.g., by `shmem_put`). I don't think providing such a constant _forces_ a library to do such checking of error conditions. We can (though I don't think we currently do) assert that it is an error to use an invalid context.

So, what are your thoughts on something like `SHMEM_CTX_INVALID`?

**Comment by nspark** [comment]
I pre-offer my apologies to @jdinan if I've misrepresented his earlier position or rationale.

**Comment by nspark** [comment]
Closed by #222
